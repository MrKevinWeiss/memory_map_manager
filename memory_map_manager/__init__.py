"""memory_map_manager package."""
__version__ = "0.1.0"
