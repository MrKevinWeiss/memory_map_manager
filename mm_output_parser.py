#!/usr/bin/env python3
"""This module generates output files based on memory maps."""
from gen_helpers import to_camel_case, to_underscore_case


def _find_unique_keys(mem_map):
    field_names = []
    for record in mem_map:
        for field_name in record.keys():
            if field_name not in field_names:
                field_names.append(field_name)
    field_names = sorted(field_names)
    field_names.insert(0, field_names.pop(field_names.index('description')))
    field_names.insert(0, field_names.pop(field_names.index('total_size')))
    field_names.insert(0, field_names.pop(field_names.index('offset')))
    field_names.insert(0, field_names.pop(field_names.index('name')))
    return field_names


def parse_mem_map_to_csv(mem_map, name_delimiter='.'):
    """Parses a memory map to a csv table string."""
    fields = _find_unique_keys(mem_map)
    csv_str = ','.join(fields)
    for record in mem_map:
        csv_str += '\n'
        name = name_delimiter.join(record['name'])
        for field in fields:
            if field not in record:
                record[field] = ''
            if record[field] is None:
                record[field] = ''
            if field == 'name':
                csv_str += name
            else:
                csv_str += str(record[field]).replace(',', '","')
            csv_str += ','
        csv_str = csv_str[:-1]
    return csv_str


def _record_to_getter(record, fxn_call_list=None):
    fxn_str = 'get_{}'.format('_'.join(record['name']))
    getter_str = "    def {}(self):\n".format(fxn_str)
    getter_str += "        \"\"\"{}\"\"\"\n".format(record['description'])
    if record['is_bitfield']:
        getter_str += "        return self.read_bits"
        getter_str += "({}, {}, {})\n".format(record['offset'],
                                              record['bit_offset'],
                                              record['bits'])
    else:
        getter_str += "        return self.read_bytes"
        getter_str += "({}, {})\n".format(record['offset'],
                                          record['total_size'])
    if isinstance(fxn_call_list, list):
        fxn_call_list.append(fxn_str)
    return getter_str


def _record_to_setter(record, fxn_call_list=None):
    fxn_str = 'set_{}'.format('_'.join(record['name']))
    setter_str = "    def {}(self, data={}):\n".format(fxn_str,
                                                       record['default'])
    setter_str += "        \"\"\"{}\"\"\"\n".format(record['description'])
    if record['is_bitfield']:
        setter_str += "        return self.write_bits"
        setter_str += "({}, {}, {}, data)\n".format(record['offset'],
                                                    record['bit_offset'],
                                                    record['bits'])
    else:
        setter_str += "        return self.write_bytes"
        setter_str += "({}, data, {})\n".format(record['offset'],
                                                record['total_size'])
    if isinstance(fxn_call_list, list):
        fxn_call_list.append(fxn_str)
    return setter_str


def parse_mem_map_to_if(mem_map, class_name, parent_module=None):
    """Parses a memory map to a python interface to a .py string."""
    if_str = ""
    if parent_module is not None:
        if_str += "from {} ".format(to_underscore_case(parent_module))
        if_str += "import {}\n\n\n".format(to_camel_case(parent_module))
    if_str += "class {}".format(to_camel_case(class_name))
    if parent_module is not None:
        if_str += "({})".format(to_camel_case(parent_module))
    if_str += ':\n'
    fxn_call_list = []
    for record in mem_map:
        if_str += _record_to_getter(record, fxn_call_list)
        if_str += '\n'
        if record['access'] != 0:
            if_str += _record_to_setter(record, fxn_call_list)
            if_str += '\n'

    if_str += "    def get_command_list(self):\n"
    if_str += "        \"\"\"A list of all possible commands\"\"\"\n"
    if_str += "        cmds = list()\n"
    for fxn_call in fxn_call_list:
        if_str += "        cmds.append(self.{})\n".format(fxn_call)
    if_str += "        return cmds\n"
    return if_str


def parse_mem_map_to_access_c(mem_map):
    """Parses access registers based on memory map to a .c string."""
    a_str = "#include \"app_access.h\"\n\n"
    a_str += "const uint8_t REG_ACCESS[] = { \n"
    size = 0
    for record in mem_map:
        if record['is_bitfield'] is False:
            for access_byte in range(record["total_size"]):
                if access_byte != 0:
                    a_str += ", "
                a_str += "0x%02X" % record["access"]
                size += 1
            if record != mem_map[-1]:
                a_str += ","
            a_str += " /* {} */\n".format('_'.join(record["name"]))
    a_str = a_str.rstrip(',')
    a_str += "/* total size %d */\n};" % size
    return a_str


def main():
    """Tests parsing each output based on example typedef."""
    from td_parser import parse_basic_typedefs
    from mm_parser import parse_typedefs_to_mem_map
    from gen_helpers import TEST_T

    typedefs = parse_basic_typedefs(TEST_T)
    mem_map = parse_typedefs_to_mem_map(typedefs)
    mem_map[0]['default'] = 6

    base_if = parse_mem_map_to_if(mem_map, 'base_if')
    test_if = parse_mem_map_to_if(mem_map, 'test_if', 'base_if')
    access_c = parse_mem_map_to_access_c(mem_map)

    print(base_if)
    print(test_if)
    print(access_c)


if __name__ == "__main__":
    main()
