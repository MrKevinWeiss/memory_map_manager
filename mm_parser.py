#!/usr/bin/env python3
"""Parses and manages memory maps from typedefs"""
import copy
from gen_helpers import PRIM_TYPES


def _copy_elements(typename, typedefs):
    for typedef in typedefs:
        if typename == typedef["name"]:
            return {'elements': copy.deepcopy(typedef["elements"])}
    raise ValueError("Cannot find {}".format(typename))


def _copy_typedef_elements(element, expanded_typedefs):
    e_type = element["type"]
    if "array_size" in element:
        a_size = element["array_size"]
        element['array'] = [_copy_elements(e_type, expanded_typedefs)
                            for i in range(a_size)]
    else:
        element.update(_copy_elements(e_type, expanded_typedefs))


def _expand_typedefs(typedefs):
    expanded_typedefs = []
    for typedef in typedefs:
        for element in typedef["elements"]:
            if element["type"] not in PRIM_TYPES:
                if "bitfield" not in element:
                    _copy_typedef_elements(element, expanded_typedefs)
        expanded_typedefs.append(typedef)
    return expanded_typedefs


def _update_offsets(elements, offset=0):
    for element in elements:
        element["offset"] = offset
        if "elements" in element:
            offset = _update_offsets(element["elements"], offset)
        elif "array" in element:
            for array_val in element["array"]:
                offset = _update_offsets(array_val["elements"], offset)
        elif "array_size" in element:
            offset += element["size"]*element["array_size"]
        else:
            offset += element["size"]
    return offset


def _expand_mem_map(typedefs, mem_map=None):

    exp_typedefs = _expand_typedefs(typedefs)
    if mem_map is None:
        mem_map = exp_typedefs[-1]
    else:
        mem_map = next(itm for itm in exp_typedefs if itm["name"] == "mem_map")
    _update_offsets(mem_map['elements'])
    return mem_map


def _element_to_bitfield_record(name, element):

    bitfields = []
    offset = 0
    for bit_info in element["bitfield"]:
        bitfield = {}
        bitfield = copy.deepcopy(element)
        bitfield["type"] = bitfield.pop("bit_type")
        bitfield["total_size"] = element["size"]
        bitfield["is_bitfield"] = True
        bitfield["bits"] = bit_info["bits"]
        bitfield["description"] = bit_info["description"]
        bitfield["bit_offset"] = offset
        bitfield.pop('bitfield', None)
        name.append(bit_info["name"])
        bitfield["name"] = copy.deepcopy(name)
        name.pop()
        offset += bit_info["bits"]
        bitfields.append(bitfield)
    return bitfields


def _element_to_mem_map_record(name, element, mem_map):
    mem_map.append(copy.deepcopy(element))
    name.append(element["name"])
    if "array_size" in element:
        mem_map[-1]["total_size"] = element["size"] * element["array_size"]
        name.append("%d" % element["array_size"])
    else:
        mem_map[-1]["total_size"] = element["size"]
    mem_map[-1].pop('array_size', None)
    mem_map[-1].pop('bitfield', None)
    mem_map[-1].pop('bit_type', None)

    mem_map[-1]["name"] = copy.deepcopy(name)
    mem_map[-1]["bits"] = element["size"] * 8
    mem_map[-1]["bit_offset"] = 0
    mem_map[-1]["is_bitfield"] = False

    if 'bitfield' in element:
        mem_map.extend(_element_to_bitfield_record(name, element))
    if "array_size" in element:
        name.pop()
    name.pop()


def _parse_elements_to_mem_map(elements, mem_map=None, name=None):
    if mem_map is None:
        mem_map = []
    if name is None:
        name = []
    for element in elements:
        if "elements" in element:
            name.append(element["name"])
            _parse_elements_to_mem_map(element["elements"], mem_map, name)
            name.pop()
        elif "array" in element:
            name.append(element["name"])
            for i, array_val in enumerate(element["array"]):
                name.append("%d" % i)
                _parse_elements_to_mem_map(array_val["elements"],
                                           mem_map, name)
                name.pop()
            name.pop()
        else:
            _element_to_mem_map_record(name, element, mem_map)
    return mem_map


def parse_typedefs_to_mem_map(typedefs, mem_map=None):
    """Parses a selected (or the last) typedef to a memory map"""
    mem_map_expanded = _expand_mem_map(typedefs, mem_map=None)
    mem_map = _parse_elements_to_mem_map(mem_map_expanded['elements'])
    return mem_map


def import_mem_map_values(mem_map, saved_mem_map,
                          generated_fields=None):
    """Imports type_name values from saved memory maps"""
    if generated_fields is None:
        generated_fields = ['bit_offset', 'bits', 'is_bitfield', 'bit_offset',
                            'name', 'offset', 'size', 'total_size', 'type']
    for saved_record in saved_mem_map:
        for record in mem_map:
            if record['name'] == saved_record['name']:
                for field_name in saved_record.keys():
                    if field_name not in generated_fields:
                        record[field_name] = saved_record[field_name]


def main():
    """Tests the parsing and updating saved values from the memory map."""
    from pprint import pprint
    from gen_helpers import TEST_T
    import td_parser

    typedefs = td_parser.parse_basic_typedefs(TEST_T)

    mem_map = parse_typedefs_to_mem_map(typedefs)
    saved_mem_map = copy.deepcopy(mem_map)
    saved_mem_map[0]['access'] = 99
    saved_mem_map[1]['default'] = '123'
    saved_mem_map[1]['description'] = 'test to alter description'

    print("================ MEM MAP =====================\n")
    pprint(mem_map, width=120)
    print("=============== SAVED MEM MAP ================\n")
    pprint(saved_mem_map, width=120)

    print('============ Diff ============')
    try:
        from deepdiff import DeepDiff
        pprint(DeepDiff(mem_map, saved_mem_map))
    except ImportError:
        print("Cannot make diff, try 'pip install deepdiff'")

    import_mem_map_values(mem_map, saved_mem_map)
    print("============== MODIFIED MEM MAP ===============\n")
    pprint(mem_map, width=120)

    print('============ Diff After Modified ============')
    try:
        from deepdiff import DeepDiff
        pprint(DeepDiff(mem_map, saved_mem_map))
    except ImportError:
        print("Cannot make diff, try 'pip install deepdiff'")


if __name__ == "__main__":
    main()
