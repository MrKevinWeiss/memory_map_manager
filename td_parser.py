#!/usr/bin/env python3
"""Fills in any missing/calculatable information of typedefs"""
from copy import deepcopy
from gen_helpers import PRIM_TYPES


def _fill_res_element(typedef, total_byte):
    res = {'name': 'res',
           'type': 'uint8_t',
           'array_size': typedef['size'] - total_byte,
           'size': 1,
           'description': 'Reserved bytes',
           'access': 0x00,
           'default': None}
    typedef["elements"].append(res)


def _update_typedef_size(typedef, total_byte):
    if 'size' in typedef:
        if typedef['size'] < total_byte:
            raise ValueError("{} to large".format(typedef["name"]))

        if typedef['size'] is not total_byte:
            _fill_res_element(typedef, total_byte)
    else:
        typedef['size'] = total_byte


def _update_element_sizes(elements, type_sizes):
    total_byte = 0
    for element in elements:
        if 'bitfield' in element:
            element["size"] = PRIM_TYPES[element["bit_type"]]
        else:
            element["size"] = type_sizes[element["type"]]

        if 'array_size' in element:
            total_byte += element["size"] * element["array_size"]
        else:
            total_byte += element["size"]
    return total_byte


def _update_typedef_sizes(typedefs):
    type_sizes = {}
    type_sizes.update(PRIM_TYPES)
    for typedef in typedefs:
        total_byte = _update_element_sizes(typedef["elements"], type_sizes)
        _update_typedef_size(typedef, total_byte)
        type_sizes[typedef["name"]] = typedef["size"]


def _fill_elements(elements, type_name, value, bitfield):
    for element in elements:
        if type_name not in element:
            element[type_name] = value
        if bitfield and 'bitfield' in element:
            for bf_val in element['bitfield']:
                if type_name not in bf_val:
                    bf_val[type_name] = value


def _fill_empty_typedefs(typedefs, type_name, value,
                         overwrite=True, bitfield=False):
    for typedef in typedefs:
        if type_name not in typedef:
            typedef[type_name] = value

        if overwrite:
            _fill_elements(typedef['elements'], type_name, value, bitfield)
        else:
            _fill_elements(typedef['elements'], type_name,
                           typedef[type_name], bitfield)


def parse_basic_typedefs(typedefs):
    """Fills in any missing/calculatable information of typedefs."""
    # Calculates sizes and res bytes
    cp_typedefs = deepcopy(typedefs)
    _update_typedef_sizes(cp_typedefs)
    _fill_empty_typedefs(cp_typedefs, 'description', '', bitfield=True)
    _fill_empty_typedefs(cp_typedefs, 'default', None)
    _fill_empty_typedefs(cp_typedefs, 'access', 1, False)
    return cp_typedefs


def main():
    """Prints example of the parsing."""
    from pprint import pprint
    from gen_helpers import TEST_T

    print('============ Before Parsing ============')
    pprint(TEST_T)
    typedefs = parse_basic_typedefs(TEST_T)
    print('============ After Parsing ============')
    pprint(typedefs)
    print('============ Diff ============')
    try:
        from deepdiff import DeepDiff

        pprint(DeepDiff(TEST_T, typedefs))
    except ImportError:
        print("Cannot make diff, try 'pip install deepdiff'")


if __name__ == "__main__":
    main()
