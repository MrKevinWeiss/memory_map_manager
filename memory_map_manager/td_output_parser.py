#!/usr/bin/env python3
"""This module generates output files based on typedefs."""


def _metadata_to_h_intro(metadata):
    intro_str = "/*\n"
    intro_str += " * Filename: {}.g\n".format(metadata["name"])
    intro_str += " * Author: {}\n".format(metadata["author"])
    intro_str += " * Revision: {}\n".format(metadata["revision"])
    intro_str += " */\n\n"
    intro_str += "#ifndef %s_H_\n" % (metadata["name"].upper())
    intro_str += "#define %s_H_\n" % (metadata["name"].upper())
    return intro_str


def _bitfield_to_c_struct(bf_element):
    bf_str = ""
    bf_str += "/* @brief {} */\n".format(bf_element["description"])
    bf_str += "typedef struct %s_TAG {\n" % (bf_element["type"])
    total_bits = 0
    for val in bf_element["bitfield"]:
        bf_str += "\t/* {} */\n".format(val["description"])
        bf_str += "\t{} {} : {};\n".format(bf_element["bit_type"],
                                           val["name"], val["bits"])
        total_bits += int(val["bits"])
    bf_str += "} %s;\n\n" % (bf_element["type"])

    if bf_element["size"] < ((float(total_bits)/8)):
        raise ValueError("Too many bits in %s for %s" %
                         (bf_element["type"], bf_element["bit_type"]))
    return bf_str


def _typedef_to_c_struct(typedef):
    c_str = ""
    for element in typedef["elements"]:
        if 'bitfield' in element:
            c_str += _bitfield_to_c_struct(element)
    c_str += "/* @brief {} */\n".format(typedef["description"])
    c_str += "typedef union %s_TAG {\n" % (typedef["name"])
    c_str += "\tstruct {\n"

    for element in typedef["elements"]:
        c_str += "\t\t/* {} */\n".format(element["description"])
        if 'array_size' in element:
            c_str += "\t\t{} {}[{}];\n".format(element["type"],
                                               element["name"],
                                               element["array_size"])
        else:
            c_str += "\t\t{} {};\n".format(element["type"], element["name"])
    c_str += "\t};\n"
    c_str += "\tuint8_t data8[{}];\n".format(typedef["size"])
    c_str += "} %s;\n" % (typedef["name"])
    return c_str


def parse_typedefs_to_h(typedefs, metadata=None):
    """Parses the typedef to a c header containing typedef structs."""
    td_str = ""
    if metadata is not None:
        td_str += _metadata_to_h_intro(metadata)
    td_str += "\n"
    td_str += "#include <stdint.h>\n"
    td_str += "\n"
    td_str += "#pragma pack(1)\n"
    for typedef in typedefs:
        td_str += _typedef_to_c_struct(typedef)
        td_str += "\n"
    td_str += "#pragma pack()\n"
    if metadata is not None:
        td_str += "#endif"
    return td_str


def main():
    """Tests parsing example typedef to c header."""
    from td_parser import parse_basic_typedefs
    from gen_helpers import TEST_T
    from gen_helpers import TEST_MD

    typedefs = parse_basic_typedefs(TEST_T)
    print(parse_typedefs_to_h(typedefs, TEST_MD))


if __name__ == "__main__":
    main()
