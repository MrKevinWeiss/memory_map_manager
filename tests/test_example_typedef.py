# Copyright (c) 2018 Kevin Weiss, for HAW Hamburg  <kevin.weiss@haw-hamburg.de>
#
# This file is subject to the terms and conditions of the MIT License. See the
# file LICENSE in the top level directory for more details.
# SPDX-License-Identifier:    MIT
"""Tests Serial Driver implmentation in RIOT PAL."""
from copy import deepcopy
import json
import difflib
import pytest
from memory_map_manager.config_import_export import import_config
from memory_map_manager.td_parser import update_typedefs
from memory_map_manager.mm_parser import parse_typedefs_to_mem_maps
from memory_map_manager.td_output_parser import parse_typedefs_to_h
from memory_map_manager.mm_output_parser import parse_mem_map_to_defaults_h
from memory_map_manager.mm_output_parser import parse_mem_map_to_map_v_c
from memory_map_manager.mm_output_parser import parse_mem_map_to_map_v_h
from memory_map_manager.mm_output_parser import parse_mem_map_to_defaults_c
from memory_map_manager.mm_output_parser import parse_mem_map_to_csv
from memory_map_manager.mm_output_parser import parse_mem_map_to_access_c


@pytest.fixture
def example_config():
    yield import_config('example_typedef.json')


# pylint: disable=W0621
def test_import_regression(regtest, example_config):
    """Tests the import from json against accepted input"""
    regtest.write(json.dumps(example_config, sort_keys=True, indent=4))


# pylint: disable=W0621
def test_updated_regression(regtest, example_config):
    """Tests the full transform of the config file against accepted value"""
    update_typedefs(example_config)
    parse_typedefs_to_mem_maps(example_config, True)
    regtest.write(json.dumps(example_config, sort_keys=True, indent=4))


# pylint: disable=W0621
def test_diff_generated_imported(regtest, example_config):
    """Tests to see if importing non-generated values works"""
    update_typedefs(example_config)
    generated_config = deepcopy(example_config)
    parse_typedefs_to_mem_maps(example_config, True)
    parse_typedefs_to_mem_maps(generated_config, False)
    imp_mm = json.dumps(example_config['mem_maps'], sort_keys=True, indent=4)
    gen_mm = json.dumps(generated_config['mem_maps'], sort_keys=True, indent=4)
    for outstr in difflib.unified_diff(imp_mm.splitlines(),
                                       gen_mm.splitlines(), n=0):
        last_str = outstr
    regtest.write(last_str)


# pylint: disable=W0621
def test_c_h_outputs(regtest, example_config):
    """Test to generated c and h stay the same."""
    update_typedefs(example_config)
    parse_typedefs_to_mem_maps(example_config, False)
    regtest.write(parse_typedefs_to_h(example_config))
    # regtest.write seems to have a bug with the rendering of 0xn, 0xm
    # as a result it is expected that we get something like 0x00, 0x????????
    regtest.write(parse_mem_map_to_access_c(example_config))
    regtest.write(parse_mem_map_to_map_v_c(example_config))
    regtest.write(parse_mem_map_to_map_v_h(example_config))
    regtest.write(parse_mem_map_to_defaults_h(example_config))
    regtest.write(parse_mem_map_to_defaults_c(example_config))


# pylint: disable=W0621
def test_csv_outputs(regtest, example_config):
    update_typedefs(example_config)
    parse_typedefs_to_mem_maps(example_config, True)
    for mem_map in example_config['mem_maps']:
        regtest.write(parse_mem_map_to_csv(mem_map))
