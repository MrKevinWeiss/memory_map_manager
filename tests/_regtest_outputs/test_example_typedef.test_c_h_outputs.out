/**
 *****************************************************************************"
 * @addtogroup example_MMM
 * @{
 * @file      example_typedef.h
 * @author    Kevin Weiss
 * @version   0.00.00
 * @details   Generated from the memory map manager version 1.0.0
 *****************************************************************************"
 */
#ifndef EXAMPLE_TYPEDEF_H
#define EXAMPLE_TYPEDEF_H


#include <stdint.h>

#pragma pack(1)
/** @brief   */
typedef struct {
	uint8_t basic_enable_reg : 1; /**<  */
	uint8_t example_prescale_reg : 5; /**<  */
	uint8_t res : 2; /**< Reserved bits */
} example_bitfield_t;

/** @brief   */
typedef struct {
	uint16_t basic_enable_reg : 6; /**<  */
	uint16_t example_prescale_reg : 7; /**<  */
	uint16_t same_offset_over_8_bit_offset : 1; /**<  */
	uint16_t res : 2; /**< Reserved bits */
} longer_bitfield_t;

/** @brief   */
typedef union {
	struct {
		uint8_t basic_element; /**<  */
		uint32_t access_3_element; /**<  */
		uint32_t access_0_element; /**<  */
		uint32_t specified_element; /**<  */
		int32_t array_element[9]; /**<  */
		uint16_t other_element; /**<  */
		example_bitfield_t using_bitfield_type_element; /**<  */
	};
	uint8_t data8[52];/**< array for padding */
} the_example_typedef_t;

/** @brief   */
typedef union {
	struct {
		uint8_t element_a; /**<  */
		uint32_t element_b_with_access_3; /**<  */
		uint8_t res[11]; /**< Reserved bytes */
	};
	uint8_t data8[16];/**< array for padding */
} simple_t;

/** @brief   */
typedef union {
	struct {
		the_example_typedef_t an_element; /**<  */
		simple_t array_of_special_type[3]; /**<  */
		longer_bitfield_t longer_bitfield; /**<  */
		uint16_t other_element; /**<  */
	};
	uint8_t data8[104];/**< array for padding */
} mem_map_example_t;

/** @brief   */
typedef union {
	struct {
		uint8_t foo; /**<  */
		uint8_t bar; /**<  */
	};
	uint8_t data8[2];/**< array for padding */
} simple_mem_map_t;

/** @brief   */
typedef union {
	struct {
		uint8_t foobar; /**< foobar description */
	};
	uint8_t data8[1];/**< array for padding */
} non_generated_mm_t;

#pragma pack()
#endif /* EXAMPLE_TYPEDEF_H */
/** @} */
/**
 *****************************************************************************"
 * @addtogroup example_MMM
 * @{
 * @file      example_access.c
 * @author    Kevin Weiss
 * @version   0.00.00
 * @}
 * @details   Generated from the memory map manager version 1.0.0
 *****************************************************************************"
 */
/* Includes -----------------------------------------------------------------*/
#include <stdio.h>

/* Global variables ---------------------------------------------------------*/
const uint8_t MEM_MAP_EXAMPLE_T_ACCESS[] = { 
0x01, /* an_element_basic_element */
0x03, 0x?????????, 0x?????????, 0x?????????, /* an_element_access_3_element */
0x00, 0x?????????, 0x?????????, 0x?????????, /* an_element_access_0_element */
0x01, 0x?????????, 0x?????????, 0x?????????, /* an_element_specified_element */
0x01, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, /* an_element_array_element */
0x01, 0x?????????, /* an_element_other_element */
0x01, /* an_element_using_bitfield_type_element */
0x01, /* array_of_special_type_0_element_a */
0x03, 0x?????????, 0x?????????, 0x?????????, /* array_of_special_type_0_element_b_with_access_3 */
0x00, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, /* array_of_special_type_0_res */
0x01, /* array_of_special_type_1_element_a */
0x03, 0x?????????, 0x?????????, 0x?????????, /* array_of_special_type_1_element_b_with_access_3 */
0x00, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, /* array_of_special_type_1_res */
0x01, /* array_of_special_type_2_element_a */
0x03, 0x?????????, 0x?????????, 0x?????????, /* array_of_special_type_2_element_b_with_access_3 */
0x00, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, 0x?????????, /* array_of_special_type_2_res */
0x01, 0x?????????, /* longer_bitfield */
0x01, 0x????????? /* other_element */
};/**< access array total size 104 */
const uint8_t SIMPLE_MEM_MAP_T_ACCESS[] = { 
0x01, /* foo */
0x01 /* bar */
};/**< access array total size 2 */
/**
 *****************************************************************************"
 * @addtogroup example_MMM
 * @{
 * @file      example_map.c
 * @author    Kevin Weiss
 * @version   0.00.00
 * @}
 * @details   Generated from the memory map manager version 1.0.0
 *****************************************************************************"
 */
/* Includes ----------------------------------------------------------------*/
#include <stdint.h>

#include "example_map.h"

/* Global variables --------------------------------------------------------*/
const char* const EXAMPLE_TYPE_NAME[] = {"uint8_t",
"int8_t",
"uint16_t",
"int16_t",
"uint32_t",
"int32_t",
"uint64_t",
"int64_t",
"char",
"float",
"double"}; /** < type_name enum */

const uint8_t  EXAMPLE_TYPE_SIZE[] = {1,
1,
2,
2,
4,
4,
8,
8,
1,
4,
8}; /** <  type_size const array */

const char* const MEM_MAP_EXAMPLE_T_NAME[] = {"an_element.basic_element",
"an_element.access_3_element",
"an_element.access_0_element",
"an_element.specified_element",
"an_element.array_element",
"an_element.other_element",
"an_element.using_bitfield_type_element",
"an_element.using_bitfield_type_element.basic_enable_reg",
"an_element.using_bitfield_type_element.example_prescale_reg",
"an_element.using_bitfield_type_element.res",
"array_of_special_type[0].element_a",
"array_of_special_type[0].element_b_with_access_3",
"array_of_special_type[0].res",
"array_of_special_type[1].element_a",
"array_of_special_type[1].element_b_with_access_3",
"array_of_special_type[1].res",
"array_of_special_type[2].element_a",
"array_of_special_type[2].element_b_with_access_3",
"array_of_special_type[2].res",
"longer_bitfield",
"longer_bitfield.basic_enable_reg",
"longer_bitfield.example_prescale_reg",
"longer_bitfield.same_offset_over_8_bit_offset",
"longer_bitfield.res",
"other_element"}; /** < name const array */

const uint8_t MEM_MAP_EXAMPLE_T_OFFSET[] = {0,
1,
5,
9,
13,
49,
51,
51,
51,
51,
52,
53,
57,
68,
69,
73,
84,
85,
89,
100,
100,
100,
100,
100,
102}; /** < offset const array */

const uint8_t MEM_MAP_EXAMPLE_T_TYPE[] = {0,
4,
4,
4,
5,
2,
0,
0,
0,
0,
0,
4,
0,
0,
4,
0,
0,
4,
0,
2,
2,
2,
2,
2,
2}; /** < type_name const array */

const uint8_t MEM_MAP_EXAMPLE_T_ARRAY_SIZE[] = {0,
0,
0,
0,
9,
0,
0,
0,
0,
0,
0,
0,
11,
0,
0,
11,
0,
0,
11,
0,
0,
0,
0,
0,
0}; /** < array_size const array */

const uint8_t MEM_MAP_EXAMPLE_T_BIT_OFFSET[] = {0,
0,
0,
0,
0,
0,
0,
0,
1,
6,
0,
0,
0,
0,
0,
0,
0,
0,
0,
0,
0,
6,
13,
14,
0}; /** < bit_offset const array */

const uint8_t MEM_MAP_EXAMPLE_T_BITS[] = {0,
0,
0,
0,
0,
0,
0,
1,
5,
2,
0,
0,
0,
0,
0,
0,
0,
0,
0,
0,
6,
7,
1,
2,
0}; /** < bits const array */

const char* const SIMPLE_MEM_MAP_T_NAME[] = {"foo",
"bar"}; /** < name const array */

const uint8_t SIMPLE_MEM_MAP_T_OFFSET[] = {0,
1}; /** < offset const array */

const uint8_t SIMPLE_MEM_MAP_T_TYPE[] = {0,
0}; /** < type_name const array */

const uint8_t SIMPLE_MEM_MAP_T_ARRAY_SIZE[] = {0,
0}; /** < array_size const array */

const uint8_t SIMPLE_MEM_MAP_T_BIT_OFFSET[] = {0,
0}; /** < bit_offset const array */

const uint8_t SIMPLE_MEM_MAP_T_BITS[] = {0,
0}; /** < bits const array */

/**
 *****************************************************************************"
 * @addtogroup example_MMM
 * @{
 * @file      example_map.h
 * @author    Kevin Weiss
 * @version   0.00.00
 * @details   Generated from the memory map manager version 1.0.0
 *****************************************************************************"
 */
#ifndef EXAMPLE_MAP_H
#define EXAMPLE_MAP_H

/* Defines ----------------------------------------------------------------- */
#define IF_VERSION_MAJOR 0 /**< Major version of interface */
#define IF_VERSION_MINOR 0 /**< Minor version of interface */
#define IF_VERSION_PATCH 0 /**< Patch version of interface */

/* Global variables -------------------------------------------------------- */
extern const char* const EXAMPLE_TYPE_NAME[]; /** < type_name enum */
extern const uint8_t  EXAMPLE_TYPE_SIZE[]; /** <  type_size const array */

#define MEM_MAP_EXAMPLE_T_NUM_OF_RECORDS 25 /**< Number of records in the map */

extern const char* const MEM_MAP_EXAMPLE_T_NAME[]; /** < name const array */
extern const uint8_t MEM_MAP_EXAMPLE_T_OFFSET[]; /** < offset const array */
extern const uint8_t MEM_MAP_EXAMPLE_T_TYPE[]; /** < type const array */
extern const uint8_t MEM_MAP_EXAMPLE_T_ARRAY_SIZE[]; /** < array_size const array */
extern const uint8_t MEM_MAP_EXAMPLE_T_BIT_OFFSET[]; /** < bit_offset const array */
extern const uint8_t MEM_MAP_EXAMPLE_T_BITS[]; /** < bits const array */
#define SIMPLE_MEM_MAP_T_NUM_OF_RECORDS 2 /**< Number of records in the map */

extern const char* const SIMPLE_MEM_MAP_T_NAME[]; /** < name const array */
extern const uint8_t SIMPLE_MEM_MAP_T_OFFSET[]; /** < offset const array */
extern const uint8_t SIMPLE_MEM_MAP_T_TYPE[]; /** < type const array */
extern const uint8_t SIMPLE_MEM_MAP_T_ARRAY_SIZE[]; /** < array_size const array */
extern const uint8_t SIMPLE_MEM_MAP_T_BIT_OFFSET[]; /** < bit_offset const array */
extern const uint8_t SIMPLE_MEM_MAP_T_BITS[]; /** < bits const array */
#endif /* EXAMPLE_MAP_H */
/** @} */
/**
 *****************************************************************************"
 * @addtogroup example_MMM
 * @{
 * @file      example_defaults.h
 * @author    Kevin Weiss
 * @version   0.00.00
 * @details   Generated from the memory map manager version 1.0.0
 *****************************************************************************"
 */
#ifndef EXAMPLE_DEFAULTS_H
#define EXAMPLE_DEFAULTS_H

/* Defines -------------------------------------------------------------------*/
/** @brief default for an_element.specified_element:  */
#define DEFAULT_AN_ELEMENT_SPECIFIED_ELEMENT 1

/** @brief Assign defaults for mem_map_example_t */
void init_defaults_mem_map_example_t(mem_map_example_t *init);

/** @brief Assign defaults for simple_mem_map_t */
void init_defaults_simple_mem_map_t(simple_mem_map_t *init);

#endif /* EXAMPLE_DEFAULTS_H */
/** @} */
/**
 *****************************************************************************"
 * @addtogroup example_MMM
 * @{
 * @file      example_defaults.c
 * @author    Kevin Weiss
 * @version   0.00.00
 * @}
 * @details   Generated from the memory map manager version 1.0.0
 *****************************************************************************"
 */
/* Includes ------------------------------------------------------------------*/
#include <stdint.h>

#include "example_typedef.h"
#include "example_defaults.h"

/* Functions -----------------------------------------------------------------*/
/** @brief Assign defaults for mem_map_example_t */
void init_defaults_mem_map_example_t(mem_map_example_t *init) {
	init->an_element.specified_element = DEFAULT_AN_ELEMENT_SPECIFIED_ELEMENT;
}

/** @brief Assign defaults for simple_mem_map_t */
void init_defaults_simple_mem_map_t(simple_mem_map_t *init) {
}

